### PR TITLE
[InstCombine] Fix transforms of two select patterns

### PR DESCRIPTION
**Pull request by dtcxzyw**
This patch fixes transforms of `select (~a | c), a, b` and `select (c & b), a, b` as discussed in [D158983](https://reviews.llvm.org/D158983).
Alive2: https://alive2.llvm.org/ce/z/ft6TDw

### CODE REVIEW

File: llvm/test/Transforms/InstCombine/select-and-or.ll [hunk 1]
```diff
@@ -587,21 +611,97 @@ define i1 @neg_and_or2(i1 %a, i1 %b, i1 %c, i1 %d) {
...
+; CHECK-NEXT:    [[COND:%.*]] = and i1 [[C]], [[B:%.*]]
+; CHECK-NEXT:    [[R:%.*]] = select i1 [[COND]], i1 [[A:%.*]], i1 [[D:%.*]]
+; CHECK-NEXT:    ret i1 [[R]]
+;
+  %c = icmp eq i32 %x, %y
+  %cond = and i1 %b, %c
+  %r = select i1 %cond, i1 %a, i1 %d
+  ret i1 %r
+}
+
```
Comment (goldsteinn) [inline llvm/test/Transforms/InstCombine/select-and-or.ll:704]: As usual, split tests to seperate commit please.

File: llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp [hunk 2]
```diff
@@ -3086,17 +3086,18 @@ Instruction *InstCombinerImpl::foldSelectOfBools(SelectInst &SI) {
                                 m_c_LogicalOr(m_Deferred(A), m_Deferred(B)))))
     return BinaryOperator::CreateXor(A, B);
 
-  // select (~a | c), a, b -> and a, (or c, freeze(b))
+  // select (~a | c), a, b -> select a, (select c, true, b), false
   if (match(CondVal, m_c_Or(m_Not(m_Specific(TrueVal)), m_Value(C))) &&
       CondVal->hasOneUse()) {
```
Comment (goldsteinn) [inline llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:3091]: `match(CondVal, m_OneUse(m_c_Or(...)))`

File: llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp [hunk 3]
```diff
@@ -3086,17 +3086,18 @@ Instruction *InstCombinerImpl::foldSelectOfBools(SelectInst &SI) {
...
-    return BinaryOperator::CreateAnd(TrueVal, Builder.CreateOr(C, FalseVal));
+    Value *OrV = Builder.CreateSelect(C, One, FalseVal);
+    return SelectInst::Create(TrueVal, OrV, Zero);
   }
-  // select (~c & b), a, b -> and b, (or freeze(a), c)
-  if (match(CondVal, m_c_And(m_Not(m_Value(C)), m_Specific(FalseVal))) &&
-      CondVal->hasOneUse()) {
-    TrueVal = Builder.CreateFreeze(TrueVal);
-    return BinaryOperator::CreateAnd(FalseVal, Builder.CreateOr(C, TrueVal));
+  // select (c & b), a, b -> select b, (select ~c, true, a), false
```
Comment (goldsteinn) [inline llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:3095]: question a bit if this is actually desirable.

Comment (nikic) [reply llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:3095]: Reduces use count and uses canonical logical and/or form, so seems desirable to me?

**Review by nikic — APPROVED** [review]
LGTM

**Pull request merged**